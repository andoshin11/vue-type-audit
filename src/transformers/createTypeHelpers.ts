function camelize(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1)
}

export function createTypeHelpers(externalComponents: Record<string, string>, eventNames: string[]) {
  return `
///////////////////// START: Type Helpers ///////////////////
import { ClassInstance, _VNodeProps, _resolveComponent, PropTypes, _VNode, _VNodeTypes, _ClassComponent, _Data, RequiredPropNames, WithEmitType, MixIntoComponent, isNeverType } from "__GLOBAL_TYPES";

type EVENT_DICT = {
${eventNames.map(n => `  on${camelize(n)}: "${n}";`).join('\n')}
}

type __ExternalComponents = {
${Object.entries(externalComponents).map(([key, val]) => `  ${key}: ClassInstance<typeof ${val}>;`).join('\n')}
}

type __ExternalComponentsProps = {
${Object.entries(externalComponents).map(([key, val]) => `  ${key}: RequiredPropNames<__ExternalComponents['${val}']>;`).join('\n')}
}

const ___Component: MixIntoComponent<typeof __Component, WithEmitType<typeof __Component, EVENT_DICT>> = __Component as any

export default ___Component;

type __PropArg<N> = N extends keyof __ExternalComponents ?  PropTypes<__ExternalComponents[N]> : _Data
type __EmitArg<N> = N extends keyof __ExternalComponents ? __ExternalComponents[N]['$options']['__emitHandlerTypes'] : {}

declare function _createVNode<N extends (keyof __ExternalComponents | _VNodeTypes | _ClassComponent)>(type: N, ...args: (N extends keyof __ExternalComponents ? true extends isNeverType<__ExternalComponentsProps[N]> ? [(_VNodeProps & __PropArg<N> & __EmitArg<N>)?, unknown?, number?, (string[] | null)?, boolean?] : [(_VNodeProps & __PropArg<N> & __EmitArg<N>), unknown?, number?, (string[] | null)?, boolean?] : [((_Data & _VNodeProps) | null)?, unknown?, number?, (string[] | null)?, boolean?])): _VNode;
///////////////////// END: Type Helpers /////////////////////
`
}
