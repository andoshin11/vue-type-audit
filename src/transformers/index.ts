export * from './template'
export * from './transformers'
