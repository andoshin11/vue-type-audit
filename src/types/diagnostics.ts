import _ts from 'typescript'

export type DiagnosticWithRange = Omit<_ts.Diagnostic, 'start' | 'length'> & { start: NonNullable<_ts.Diagnostic['start']>; length: NonNullable<_ts.Diagnostic['length']> }
