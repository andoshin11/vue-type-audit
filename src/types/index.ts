import _ts from 'typescript'

export * from './diagnostics'

export type FileEntry = Map<string, { version: number; scriptSnapshot: _ts.IScriptSnapshot }>
