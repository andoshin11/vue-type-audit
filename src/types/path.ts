import { Brand } from '.'

export type RawVueFileName = Brand<string, 'RawVueFileName'>

export type TSVueFileName = Brand<string, 'TSVueFileName'>

export type VueTemplateFileName = Brand<string, 'VueTemplateFileName'>

export type VueScriptFileName = Brand<string, 'VueScriptFileName'>

export type CompiledVueTemplateFileName = Brand<string, 'CompiledVueTemplateFileName'>

export type CompiledVueScriptFileName = Brand<string, 'CompiledVueTemplateFileName'>

export type TransformedCompiledVueTemplateFileName = Brand<string, 'TransformedCompiledVueTemplateFileName'>

export type TransformedVueScriptFileName = Brand<string, 'TransformedVueScriptFileName'>
