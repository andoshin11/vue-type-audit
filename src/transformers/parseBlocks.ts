import { parse as parseComponent } from '@vue/compiler-sfc'
import { compile } from '@vue/compiler-dom'
import { RawVueFileName, SourcemapEntry } from '../types'
import { toVueScriptFileName, toVueTemplateFileName, toCompiledVueTemplateFileName } from '../helpers'

export function parseBlocks(
  fileName: RawVueFileName,
  content: string,
  sourcemapEntry: SourcemapEntry
) {
  let { template, script } = parseComponent(content, { sourceMap: true, filename: fileName }).descriptor

  // `template` can be empty, yet you always require `script` block
  if (!script) {
    content += `
<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
})
</script>
    `
    const appended = parseComponent(content, { sourceMap: true, filename: fileName }).descriptor
    template = appended.template
    script = appended.script!
  }

  /**
   * Register sourcemap for blocks
   */

  // `template`
  const vueTemplateFileName = toVueTemplateFileName(fileName)
  // FIXME
  sourcemapEntry.set(vueTemplateFileName, { map: template ? template.map : undefined })

  const compiledVueTemplateFileName = toCompiledVueTemplateFileName(vueTemplateFileName)
  const compiled = compile(template ? template.content : '', { mode: 'module', sourceMap: true, filename: vueTemplateFileName })
  sourcemapEntry.set(compiledVueTemplateFileName, { map: compiled.map })

  // `script`
  const vueScriptFileName = toVueScriptFileName(fileName)
  sourcemapEntry.set(vueScriptFileName, {
    map: {
      ...script.map!,
      file: vueScriptFileName
    }
  })

  let templateContent = !!compiled.code ? compiled.code : undefined
  let scriptContent = script.content

  return {
    templateContent,
    scriptContent
  }
}
