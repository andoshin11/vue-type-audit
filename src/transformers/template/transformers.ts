import _ts from 'typescript'
import { SourceMapGenerator, RawSourceMap } from 'source-map'
import { CompiledVueTemplateFileName } from '../../types'
import { toTransformedCompiledVueTemplateFileName } from '../../helpers'

export function transformTemplate(
  fileName: CompiledVueTemplateFileName,
  content: string
) {
  const transformedCompiledVueTemplateFileName = toTransformedCompiledVueTemplateFileName(fileName)
  const gen = new SourceMapGenerator({ file: transformedCompiledVueTemplateFileName })
  gen.setSourceContent(fileName, content)

  // escape overridden identifiers
  let transformed = content
  transformed = transformed.replace(/_createVNode/, '#createVNode')
  transformed = transformed.replace(/_resolveComponent/, '#resolveComponent')

  // strip export assignment
  transformed = transformed.replace('export function render(_ctx, _cache)', 'function render(_ctx: __ComponentType)')

  transformed
    .split('\n')
    .forEach((_, index) => {
      gen.addMapping({
        source: fileName,
        original: { line: index + 1, column: 0 },
        generated: { line: index + 1, column: 0 }
      })
    })

  const map = JSON.parse(gen.toString()) as RawSourceMap
  return {
    transformed,
    map
  }
}
