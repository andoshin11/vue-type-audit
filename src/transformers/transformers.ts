import _ts from 'typescript'
import * as fs from 'fs'
import { parse as parseComponent } from '@vue/compiler-sfc'
import { compile } from '@vue/compiler-dom'
import { RawVueFileName, SourcemapEntry } from '../types'
import { transformScript } from './script'
import { transformTemplate } from './template'
import { toVueScriptFileName, toTSVueFIleName, toVueTemplateFileName, toCompiledVueTemplateFileName } from '../helpers'

export function transformVueFile(
  fileName: RawVueFileName,
  content: string,
  sourcemapEntry: SourcemapEntry
) {
  const tsVueFileName = toTSVueFIleName(fileName)

  // Parse blocks
  const { template, script } = parseComponent(content, { sourceMap: true, filename: fileName }).descriptor

  // `template` can be empty, yet you always require `script` block
  if (!script) {
    throw new Error('<script/> block is missing in ' + fileName)
  }

  /**
   * Register sourcemap for blocks
   */

  // `template`
  const vueTemplateFileName = toVueTemplateFileName(fileName)
  sourcemapEntry.set(vueTemplateFileName, { map: template ? template.map : undefined })

  const compiledVueTemplateFileName = toCompiledVueTemplateFileName(vueTemplateFileName)
  const compiled = compile(template ? template.content : '', { mode: 'module', sourceMap: true, filename: vueTemplateFileName })
  sourcemapEntry.set(compiledVueTemplateFileName, { map: compiled.map })


  // `script`
  const vueScriptFileName = toVueScriptFileName(fileName)
  sourcemapEntry.set(vueScriptFileName, {
    map: {
      ...script.map!,
      file: vueScriptFileName
    }
  })

  let templateContent = !!compiled.code ? compiled.code : undefined
  let scriptContent = script.content


  /**
   * Create outputs
   */
  const preparationBlock = createPreparationBlock()

  const scriptBlock = transformScript(vueScriptFileName, scriptContent, _ts)
  const templateBlock = transformTemplate(compiledVueTemplateFileName, templateContent)

  const preparationBlockLinesNum = preparationBlock.split('\n').length
  const scriptBlockLinesNum = preparationBlockLinesNum + 1 + scriptBlock.transformed.split('\n').length

  const map = {
    version: '3',
    file: tsVueFileName,
    sections: [
      // script block
      {
        offset: {
          line: preparationBlockLinesNum,
          column: 0
        },
        map: scriptBlock.map
      },
      // template block
      {
        offset: {
          line: scriptBlockLinesNum,
          column: 0
        },
        map: templateBlock.map
      }
    ]
  } as any

  sourcemapEntry.set(tsVueFileName, { map })

  const output = preparationBlock + '\n' + scriptBlock.transformed + '\n' + templateBlock.transformed
  fs.writeFileSync(toTSVueFIleName(fileName), output)

  return output
}

function createPreparationBlock() {
  return 'import { ClassInstance, __createVNodeFromExternalComponent, _VNodeProps, PropTypes, _VNode, _VNodeTypes, _ClassComponent, _Data } from "__GLOBAL_TYPES";'
}
