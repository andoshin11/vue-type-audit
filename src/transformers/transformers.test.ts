import ts from 'typescript'
import { transformVueFile } from './transformers'
import { RawVueFileName, SourcemapEntry } from '../types'

describe('transformers', () => {

  test('transformScript', () => {
    const cases = [
      `
<template>
  <div>
    <img src="./logo.png">
    <h1>Hello Vue 3!</h1>
    <button @click="inc">Clicked {{ count.length }} times.</button>
    <HelloWorld/>
  </div>
</template>

<script lang="ts">
import { ref, defineComponent } from 'vue'
import HelloWorld from './HelloWorld.vue'

export default defineComponent({
  components: {
    HelloWorld
  },
  setup(props) {
    const count = ref(0)
    const inc = () => {
      count.value++
    }
    console.log(count.value.length)

    return {
      count,
      inc
    }
  }
})
</script>

<style scoped>
img {
  width: 200px;
}
h1 {
  font-family: Arial, Helvetica, sans-serif;
}
</style>    
    `,
    `
<template>
  <div>
    <h1>{{ msg }}</h1>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  props: {
    msg: {
      type: String,
      required: true
    }
  }
})
</script>
    `,
    ]

    const sourcemapEntry = new Map() as SourcemapEntry

    cases.forEach(c => {
      const transformed = transformVueFile('test.vue' as RawVueFileName, c, sourcemapEntry, ts)
      expect(transformed).toMatchSnapshot()
    })

    expect(sourcemapEntry).toMatchSnapshot()
  })
})
