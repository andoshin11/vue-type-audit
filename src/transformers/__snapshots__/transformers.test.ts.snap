// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformers transformScript 1`] = `
"import { ClassInstance, __createVNodeFromExternalComponent, _VNodeProps, PropTypes, _VNode, _VNodeTypes, _ClassComponent, _Data } from \\"__GLOBAL_TYPES\\";

import { ref, defineComponent } from 'vue'
import HelloWorld from './HelloWorld.vue'

const __Component = defineComponent({
  components: {
    HelloWorld
  },
  setup(props) {
    const count = ref(0)
    const inc = () => {
      count.value++
    }
    console.log(count.value.length)

    return {
      count,
      inc
    }
  }
})


type __ExternalComponents = {
_component_HelloWorld: ClassInstance<typeof HelloWorld>;
}

type __ResolveComponentNames = {
HelloWorld: '_component_HelloWorld';
}


export default __Component;
type __ComponentType = ClassInstance<typeof __Component>;


declare function _createVNode<N extends (keyof __ExternalComponents | _VNodeTypes | _ClassComponent), C = N extends keyof __ExternalComponents ?  __ExternalComponents[N] : unknown>(type: N, ...args: (N extends keyof __ExternalComponents ? {} extends PropTypes<C> ? [(_VNodeProps & PropTypes<C>)?, unknown?, number?, (string[] | null)?] : [(_VNodeProps & PropTypes<C>), unknown?, number?, (string[] | null)?] : [((_Data & _VNodeProps) | null)?, unknown?, number?, (string[] | null)?])): _VNode;

declare function _resolveComponent<N extends (keyof __ResolveComponentNames | string)>(name: N): N extends keyof __ResolveComponentNames ? __ResolveComponentNames[N] : string

import { createVNode as #createVNode, toDisplayString as _toDisplayString, resolveComponent as #resolveComponent, openBlock as _openBlock, createBlock as _createBlock } from \\"vue\\"

function render(_ctx: __ComponentType) {
  const _component_HelloWorld = _resolveComponent(\\"HelloWorld\\")

  return (_openBlock(), _createBlock(\\"div\\", null, [
    _createVNode(\\"img\\", { src: \\"./logo.png\\" }),
    _createVNode(\\"h1\\", null, \\"Hello Vue 3!\\"),
    _createVNode(\\"button\\", { onClick: _ctx.inc }, \\"Clicked \\" + _toDisplayString(_ctx.count.length) + \\" times.\\", 9 /* TEXT, PROPS */, [\\"onClick\\"]),
    _createVNode(_component_HelloWorld)
  ]))
}"
`;

exports[`transformers transformScript 2`] = `
"import { ClassInstance, __createVNodeFromExternalComponent, _VNodeProps, PropTypes, _VNode, _VNodeTypes, _ClassComponent, _Data } from \\"__GLOBAL_TYPES\\";

import { defineComponent } from 'vue'

const __Component = defineComponent({
  props: {
    msg: {
      type: String,
      required: true
    }
  }
})


type __ExternalComponents = {}

type __ResolveComponentNames = {}


export default __Component;
type __ComponentType = ClassInstance<typeof __Component>;


declare function _createVNode<N extends (keyof __ExternalComponents | _VNodeTypes | _ClassComponent), C = N extends keyof __ExternalComponents ?  __ExternalComponents[N] : unknown>(type: N, ...args: (N extends keyof __ExternalComponents ? {} extends PropTypes<C> ? [(_VNodeProps & PropTypes<C>)?, unknown?, number?, (string[] | null)?] : [(_VNodeProps & PropTypes<C>), unknown?, number?, (string[] | null)?] : [((_Data & _VNodeProps) | null)?, unknown?, number?, (string[] | null)?])): _VNode;

declare function _resolveComponent<N extends (keyof __ResolveComponentNames | string)>(name: N): N extends keyof __ResolveComponentNames ? __ResolveComponentNames[N] : string

import { toDisplayString as _toDisplayString, createVNode as #createVNode, openBlock as _openBlock, createBlock as _createBlock } from \\"vue\\"

function render(_ctx: __ComponentType) {
  return (_openBlock(), _createBlock(\\"div\\", null, [
    _createVNode(\\"h1\\", null, _toDisplayString(_ctx.msg), 1 /* TEXT */)
  ]))
}"
`;

exports[`transformers transformScript 3`] = `
Map {
  "test.vue.template.ts" => Object {
    "map": Object {
      "file": "test.vue",
      "mappings": ";AAEA;AACA;AACA",
      "names": Array [],
      "sourceRoot": "",
      "sources": Array [
        "test.vue",
      ],
      "sourcesContent": Array [
        "
<template>
  <div>
    <h1>{{ msg }}</h1>
  </div>
</template>

<script lang=\\"ts\\">
import { defineComponent } from 'vue'

export default defineComponent({
  props: {
    msg: {
      type: String,
      required: true
    }
  }
})
</script>
    ",
      ],
      "version": 3,
    },
  },
  "test.vue.template.compiled.ts" => Object {
    "map": Object {
      "mappings": ";;;wBACE,aAEM;IADJ,aAAkB,6BAAXA,QAAG",
      "names": Array [
        "msg",
      ],
      "sources": Array [
        "test.vue.template.ts",
      ],
      "sourcesContent": Array [
        "
  <div>
    <h1>{{ msg }}</h1>
  </div>
",
      ],
      "version": 3,
    },
  },
  "test.vue.script.ts" => Object {
    "map": Object {
      "file": "test.vue.script.ts",
      "mappings": ";AAQA;;AAEA;AACA;AACA;AACA;AACA;AACA;AACA;AACA",
      "names": Array [],
      "sourceRoot": "",
      "sources": Array [
        "test.vue",
      ],
      "sourcesContent": Array [
        "
<template>
  <div>
    <h1>{{ msg }}</h1>
  </div>
</template>

<script lang=\\"ts\\">
import { defineComponent } from 'vue'

export default defineComponent({
  props: {
    msg: {
      type: String,
      required: true
    }
  }
})
</script>
    ",
      ],
      "version": 3,
    },
  },
  "test.vue.ts" => Object {
    "map": Object {
      "file": "test.vue.ts",
      "sections": Array [
        Object {
          "map": Object {
            "file": "test.vue.script.transformed.ts",
            "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA",
            "names": Array [],
            "sources": Array [
              "test.vue.script.ts",
            ],
            "sourcesContent": Array [
              "
import { defineComponent } from 'vue'

export default defineComponent({
  props: {
    msg: {
      type: String,
      required: true
    }
  }
})
",
            ],
            "version": 3,
          },
          "offset": Object {
            "column": 0,
            "line": 2,
          },
        },
        Object {
          "map": Object {
            "file": "test.vue.template.compiled.transformed.ts",
            "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA",
            "names": Array [],
            "sources": Array [
              "test.vue.template.compiled.ts",
            ],
            "sourcesContent": Array [
              "import { toDisplayString as _toDisplayString, createVNode as _createVNode, openBlock as _openBlock, createBlock as _createBlock } from \\"vue\\"

export function render(_ctx, _cache) {
  return (_openBlock(), _createBlock(\\"div\\", null, [
    _createVNode(\\"h1\\", null, _toDisplayString(_ctx.msg), 1 /* TEXT */)
  ]))
}",
            ],
            "version": 3,
          },
          "offset": Object {
            "column": 0,
            "line": 28,
          },
        },
      ],
      "version": "3",
    },
  },
}
`;
